simul/platform/mininet/start.py

```diff
@@ -0,0 +1,243 @@
+#!/usr/bin/python
+
+"""
+This will run a number of hosts on the server and do all
+the routing to being able to connect to the other mininets.
+
+You have to give it a list of server/net/nbr for each server
+that has mininet installed and what subnet should be run
+on it.
+
+It will create nbr+1 entries for each net, where the ".1" is the
+router for the net, and ".2"..".nbr+1" will be the nodes.
+"""
+
+import sys, time, threading, os, datetime, contextlib, errno, platform
+
+from mininet.topo import Topo
+from mininet.net import Mininet
+from mininet.cli import CLI
+from mininet.log import lg, setLogLevel
+from mininet.node import Node, Host
+from mininet.util import netParse, ipAdd, irange
+from mininet.nodelib import NAT
+from mininet.link import TCLink
+from subprocess import Popen, PIPE, call
+
+# The port used for socat
+socatPort = 5000
+# What debugging-level to use
+debugging = 1
+# Logging-file
+logfile = "/tmp/mininet.log"
+logdone = "/tmp/done.log"
+# Whether a ssh-daemon should be launched
+runSSHD = False
+
+def dbg(lvl, *str):
+    if lvl <= 1:
+        print str
+
+class BaseRouter(Node):
+    """"A Node with IP forwarding enabled."""
+    def config( self, rootLog=None, **params ):
+        super(BaseRouter, self).config(**params)
+        dbg( 2, "Starting router %s at %s" %( self.IP(), rootLog) )
+        for (gw, n, i) in otherNets:
+            dbg( 3, "Adding route for", n, gw )
+            self.cmd( 'route add -net %s gw %s' % (n, gw) )
+        if runSSHD:
+            self.cmd('/usr/sbin/sshd -D &')
+
+        self.cmd( 'sysctl net.ipv4.ip_forward=1' )
+        self.cmd( 'iptables -t nat -I POSTROUTING -j MASQUERADE' )
+        socat = "socat OPEN:%s,creat,append udp4-listen:%d,reuseaddr,fork" % (logfile, socatPort)
+        self.cmd( '%s &' % socat )
+        if rootLog:
+            self.cmd('tail -f %s | socat - udp-sendto:%s:%d &' % (logfile, rootLog, socatPort))
+
+    def terminate( self ):
+        dbg( 2, "Stopping router" )
+        for (gw, n, i) in otherNets:
+            dbg( 3, "Deleting route for", n, gw )
+            self.cmd( 'route del -net %s gw %s' % (n, gw) )
+
+        self.cmd( 'sysctl net.ipv4.ip_forward=0' )
+        self.cmd( 'killall socat' )
+        self.cmd( 'iptables -t nat -D POSTROUTING -j MASQUERADE' )
+        super(BaseRouter, self).terminate()
+
+
+class Cothority(Host):
+    """A cothority running in a host"""
+    def config(self, gw=None, simul="", **params):
+        self.gw = gw
+        self.simul = simul
+        super(Cothority, self).config(**params)
+        if runSSHD:
+            self.cmd('/usr/sbin/sshd -D &')
+
+    def startCothority(self):
+        socat="socat -v - udp-sendto:%s:%d" % (self.gw, socatPort)
+
+        args = "-debug %s -address %s:2000 -simul %s" % (debugging, self.IP(), self.simul)
+        if True:
+            args += " -monitor %s:10000" % global_root
+        ldone = ""
+        if self.IP().endswith(".0.2"):
```

Comment:
Added a comment. ".0.1" is the 'BaseRouter', and ".0.2" is the first conode, then comes ".0.3", ..., ".0.255", ".1.0", ...